File: backend/pkg/core/worker/pipeline.go
package worker

import (
	"context"
	"fmt"
	"log/slog"
	"time"
)

const PipelineWorkerName = "pipeline"

// PipelineWorker orchestrates the crawl -> insight workflow.
type PipelineWorker struct {
	manager       *WorkerManager
	statusService *StatusService
	logger        *slog.Logger
}

// NewPipelineWorker creates a worker that runs the full insights pipeline.
func NewPipelineWorker(logger *slog.Logger, manager *WorkerManager) *PipelineWorker {
	if logger == nil {
		logger = slog.Default()
	}

	return &PipelineWorker{
		manager:       manager,
		statusService: GetStatusService(),
		logger:        logger,
	}
}

// Name returns the worker name.
func (w *PipelineWorker) Name() string {
	return PipelineWorkerName
}

// Run executes the crawl worker and then the insight worker.
func (w *PipelineWorker) Run(ctx context.Context) error {
	if w.manager == nil {
		return fmt.Errorf("worker manager not configured")
	}

	if err := w.runStep(ctx, "crawl", "Crawling sources", 0, 50); err != nil {
		return fmt.Errorf("crawl step failed: %w", err)
	}

	if err := w.runStep(ctx, "insight", "Generating insights", 50, 50); err != nil {
		return fmt.Errorf("insight step failed: %w", err)
	}

	w.statusService.UpdateStatus(w.Name(), StateRunning, 100, "Pipeline complete")
	return nil
}

func (w *PipelineWorker) runStep(
	ctx context.Context,
	workerName string,
	label string,
	baseProgress int,
	progressSpan int,
) error {
	w.statusService.UpdateStatus(w.Name(), StateRunning, baseProgress, label)
	launchedAt := time.Now()

	if err := w.manager.RunWorkerNow(workerName); err != nil {
		return err
	}

	return w.waitForStep(ctx, workerName, label, baseProgress, progressSpan, launchedAt)
}

func (w *PipelineWorker) waitForStep(
	ctx context.Context,
	workerName string,
	label string,
	baseProgress int,
	progressSpan int,
	launchedAt time.Time,
) error {
	subscriber := w.statusService.Subscribe()
	defer w.statusService.Unsubscribe(subscriber)

	done, err := w.handleStepStatus(ctx, workerName, label, baseProgress, progressSpan, launchedAt)
	if err != nil {
		return err
	}
	if done {
		return nil
	}

	for {
		select {
		case <-ctx.Done():
			if stopErr := w.manager.StopWorker(workerName); stopErr != nil && stopErr != ErrWorkerNotRunning {
				w.logger.Warn("failed to stop pipeline step after cancellation", "worker", workerName, "error", stopErr)
			}
			return ctx.Err()
		case update, ok := <-subscriber:
			if !ok || update.WorkerName != workerName {
				continue
			}

			done, err := w.handleStepStatusFromValue(workerName, label, baseProgress, progressSpan, launchedAt, update.Status)
			if err != nil {
				return err
			}
			if done {
				return nil
			}
		}
	}
}

func (w *PipelineWorker) handleStepStatus(
	ctx context.Context,
	workerName string,
	label string,
	baseProgress int,
	progressSpan int,
	launchedAt time.Time,
) (bool, error) {
	select {
	case <-ctx.Done():
		return false, ctx.Err()
	default:
	}

	status := w.statusService.GetStatus(workerName)
	if status == nil {
		return false, nil
	}

	return w.handleStepStatusFromValue(workerName, label, baseProgress, progressSpan, launchedAt, *status)
}

func (w *PipelineWorker) handleStepStatusFromValue(
	workerName string,
	label string,
	baseProgress int,
	progressSpan int,
	launchedAt time.Time,
	status WorkerStatus,
) (bool, error) {
	if !isFreshStepStatus(status, launchedAt) {
		return false, nil
	}

	progress := baseProgress + (status.Progress * progressSpan / 100)
	message := label
	if status.Message != "" {
		message = fmt.Sprintf("%s: %s", label, status.Message)
	}

	switch status.State {
	case StateRunning:
		w.statusService.UpdateStatus(w.Name(), StateRunning, progress, message)
	case StateCompleted:
		w.statusService.UpdateStatus(w.Name(), StateRunning, baseProgress+progressSpan, fmt.Sprintf("%s complete", label))
		return true, nil
	case StateFailed:
		errMessage := status.Error
		if errMessage == "" {
			errMessage = status.Message
		}
		if errMessage == "" {
			errMessage = "step failed"
		}
		w.statusService.SetError(w.Name(), fmt.Sprintf("%s failed: %s", label, errMessage))
		return false, fmt.Errorf("%s", errMessage)
	}

	return false, nil
}

func isFreshStepStatus(status WorkerStatus, launchedAt time.Time) bool {
	if status.State == StateIdle {
		return false
	}

	if status.StartedAt == nil {
		return status.State == StateRunning
	}

	return !status.StartedAt.Before(launchedAt)
}

File: backend/src/core/worker/pipeline.rs
use std::sync::{Arc, Weak};

use async_trait::async_trait;
use serde_json::{Map, Value};

use super::manager::WorkerOutcome;
use super::{
    RunMetadata, StatusService, Worker, WorkerContext, WorkerFailure, WorkerManager, WorkerResult,
    WorkerResultStatus, WorkerState,
};

pub const PIPELINE_WORKER_NAME: &str = "pipeline";

pub struct PipelineWorker {
    manager: Weak<WorkerManager>,
    status: Arc<StatusService>,
}

impl PipelineWorker {
    pub fn new(manager: Weak<WorkerManager>, status: Arc<StatusService>) -> Self {
        Self { manager, status }
    }

    async fn run_step(
        &self,
        context: &WorkerContext,
        worker_name: &str,
        label: &str,
        base_progress: i32,
        progress_span: i32,
    ) -> Result<WorkerResultStatus, WorkerFailure> {
        self.status.update_status(
            PIPELINE_WORKER_NAME,
            WorkerState::Running,
            base_progress,
            label,
        );
        let _ = context
            .task_run()
            .start_step(worker_name, worker_name, Some(label.to_owned()))
            .await;
        let manager = self
            .manager
            .upgrade()
            .ok_or_else(|| WorkerFailure::new("worker manager not configured"))?;
        let metadata = RunMetadata {
            parent_run_id: context.task_run().run_id(),
            trigger_source: "workflow".to_owned(),
            ..RunMetadata::default()
        };
        let child = manager
            .run_and_wait(worker_name, metadata, context.cancellation())
            .await;
        let (child_run_id, outcome) = match child {
            Ok(value) => value,
            Err(error) => {
                let message = error.to_string();
                let _ = context
                    .task_run()
                    .finish_step(
                        worker_name,
                        "failed",
                        Some(format!("{label} failed")),
                        Some(message.clone()),
                        Map::new(),
                    )
                    .await;
                return Err(WorkerFailure::new(message));
            }
        };
        let result: Result<WorkerResultStatus, WorkerFailure> = match outcome {
            WorkerOutcome::Completed(status) => Ok(status),
            WorkerOutcome::Failed(error) => Err(WorkerFailure::new(error)),
            WorkerOutcome::Cancelled => Err(WorkerFailure::new("operation cancelled")),
        };
        match &result {
            Ok(status) => {
                self.status.update_status(
                    PIPELINE_WORKER_NAME,
                    WorkerState::Running,
                    base_progress + progress_span,
                    format!("{label} complete"),
                );
                let summary = if *status == WorkerResultStatus::Warning {
                    format!("{label} completed with warnings")
                } else {
                    format!("{label} complete")
                };
                let mut metrics = Map::new();
                if let Some(child_run_id) = child_run_id {
                    metrics.insert(
                        "child_run_id".to_owned(),
                        Value::String(child_run_id.to_string()),
                    );
                }
                let _ = context
                    .task_run()
                    .finish_step(worker_name, status.as_str(), Some(summary), None, metrics)
                    .await;
            }
            Err(error) => {
                self.status.set_error(
                    PIPELINE_WORKER_NAME,
                    format!("{label} failed: {}", error.message()),
                );
                let _ = context
                    .task_run()
                    .finish_step(
                        worker_name,
                        "failed",
                        Some(format!("{label} failed")),
                        Some(error.to_string()),
                        Map::new(),
                    )
                    .await;
            }
        }
        result
    }
}

#[async_trait]
impl Worker for PipelineWorker {
    fn name(&self) -> &str {
        PIPELINE_WORKER_NAME
    }

    async fn run(&self, context: WorkerContext) -> Result<WorkerResult, WorkerFailure> {
        let crawl = self
            .run_step(&context, "crawl", "Crawling sources", 0, 50)
            .await
            .map_err(|error| WorkerFailure::new(format!("crawl step failed: {error}")))?;
        let insight = self
            .run_step(&context, "insight", "Generating insights", 50, 50)
            .await
            .map_err(|error| WorkerFailure::new(format!("insight step failed: {error}")))?;
        self.status.update_status(
            PIPELINE_WORKER_NAME,
            WorkerState::Running,
            100,
            "Pipeline complete",
        );

        let warning =
            crawl == WorkerResultStatus::Warning || insight == WorkerResultStatus::Warning;
        let mut result = if warning {
            WorkerResult::warning(
                "Pipeline completed with warnings",
                [
                    (crawl == WorkerResultStatus::Warning)
                        .then_some("Crawl completed with warnings".to_owned()),
                    (insight == WorkerResultStatus::Warning)
                        .then_some("Insight generation completed with warnings".to_owned()),
                ]
                .into_iter()
                .flatten()
                .collect(),
            )
        } else {
            WorkerResult::completed("Pipeline completed successfully")
        };
        result.metrics.insert(
            "crawl_status".to_owned(),
            Value::String(crawl.as_str().to_owned()),
        );
        result.metrics.insert(
            "insight_status".to_owned(),
            Value::String(insight.as_str().to_owned()),
        );
        Ok(result)
    }
}
